rosidl_parser/rosidl_parser/definition.py

```diff
@@ -0,0 +1,343 @@
+
+    __slots__ = ()
+
+
+class BasicType(BaseType):
+
+    __slots__ = ('type', )
+
+    def __init__(self, typename):
+        super().__init__()
+        assert typename in BASIC_TYPES
+        self.type = typename
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.type == other.type
+
+
+class NamedType(BaseType):
+
+    __slots__ = ('name')
+
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.name == other.name
+
+
+class NamespacedType(BaseType):
+
+    __slots__ = ('namespaces', 'name')
+
+    def __init__(self, namespaces, name):
+        super().__init__()
+        self.namespaces = namespaces
+        self.name = name
+
+    def __eq__(self, other):
+        return super().__eq__(other) and \
+            self.namespaces == other.namespaces and self.name == other.name
+
+
+class BaseString(BaseType):
+
+    __slots__ = ('maximum_size', )
+
+    def __init__(self, maximum_size=None):
+        self.maximum_size = maximum_size
+
+    def __eq__(self, other):
+        return super().__eq__(other) and \
+            self.maximum_size == other.maximum_size
+
+
+class String(BaseString):
+
+    __slots__ = ()
+
+    def __init__(self, maximum_size=None):
+        super().__init__(maximum_size=maximum_size)
+
+
+class WString(BaseString):
+
+    __slots__ = ()
+
+    def __init__(self, maximum_size=None):
+        super().__init__(maximum_size=maximum_size)
+
+
+# the following types are templated on a base type
+
+class NestedType(AbstractType):
+
+    __slots__ = ('basetype', )
+
+    def __init__(self, basetype):
+        super().__init__()
+        assert isinstance(basetype, BaseType), basetype
+        self.basetype = basetype
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.basetype == other.basetype
+
+
+class Array(NestedType):
+
+    __slots__ = ('size')
+
+    def __init__(self, basetype, size):
+        super().__init__(basetype)
+        self.size = size
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.size == other.size
+
+
+class Sequence(NestedType):
+
+    __slots__ = set()
+
+    def __init__(self, basetype):
+        super().__init__(basetype)
+
+
+class UnboundedSequence(Sequence):
+
+    __slots__ = ()
+
+    def __init__(self, basetype):
+        super().__init__(basetype)
+
+
+class BoundedSequence(Sequence):
+
+    __slots__ = ('upper_bound', )
+
+    def __init__(self, basetype, upper_bound):
+        super().__init__(basetype)
+        assert isinstance(upper_bound, int)
+        self.upper_bound = upper_bound
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.upper_bound == other.upper_bound
+
+
+class Annotation:
+
+    __slots__ = ('name', 'value')
+
+    def __init__(self, name, value):
+        assert isinstance(name, str)
+        self.name = name
+        self.value = value
```

Comment:
Is this an annotation declaration or application? If the latter and an annotation has multiple parameters like `@range(min=10, max=20)`, what is stored in value?

Reply:
The value is annotation specific. For `range` it would be a dictionary: `{'min': 10, 'max': 20}`

Reply:
Mind adding a comment describing value in the code?

Reply:
Added docblock in 7647c6b042bd4c0d4135baf513c34948ec82e729.